### ISSUE TITLE
Publishing a snapshot results in "exit status 2"

### ISSUE THREAD

**Issue by rockstar** [lxc/lxd#2201]
# Required information

 * Distribution: Ubuntu
 * Distribution version: Xenial
 * The output of "lxc info" or if that fails:

apicompat: 0
auth: trusted
environment:
  addresses:
  - 127.0.0.1:8443
  architectures:
  - x86_64
  - i686
  driver: lxc
  driverversion: 2.0.1
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 4.4.0-24-generic
  server: lxd
  serverpid: 19722
  serverversion: 2.0.2
  storage: zfs
  storageversion: "5"
config:
  core.https_address: 127.0.0.1:8443
  core.trust_password: true
  storage.zfs_pool_name: lxd
public: false

# Issue description

A brief description of what failed or what could be improved.

# Steps to reproduce

 1. Create a snapshot of the container
 2. Publish the snapshot as an image by POSTing to /1.0/images with {"public": false, "source": {"type": "snapshot", "name": "container-name/snapshot-name"
 3. Wait for the resulting operation to complete, and see the "exit status 2" in the "err" of the metadata.

The operation wait call does take some time, and I can see gzip being executed in top as it's waiting. In the end, it errors. @tych0 hypothesized that it might not be getting the right tar path.

**Comment by stgraber** [comment]
Can you confirm this still happens on 2.0.3? I can't remember us fixing this in the 2.0.3 release but probably worth making sure before debugging this more.

**Comment by rockstar** [comment]
Shoot, yeah, I did `lxc info` in the wrong terminal, but I was indeed experiencing the bug in 2.0.3.

**Comment by tych0** [comment]
FWIW, I just tried this via the CLI and it seems to work just fine, so something else must be amok:

```
dev2:~ lxc snapshot x
dev2:~ lxc publish x/snap0
Container published with fingerprint: 2cae69bd55ffb7aa38ede2987d2c8cd6892ffbe11d35039558610ca742b2a988
dev2:~ lxc image list 
+--------+--------------+--------+-----------------------------------------------+--------+----------+------------------------------+
| ALIAS  | FINGERPRINT  | PUBLIC |                  DESCRIPTION                  |  ARCH  |   SIZE   |         UPLOAD DATE          |
+--------+--------------+--------+-----------------------------------------------+--------+----------+------------------------------+
| tester | 4ab05745048a | no     |                                               | x86_64 | 226.37MB | Jul 12, 2016 at 3:50pm (UTC) |
+--------+--------------+--------+-----------------------------------------------+--------+----------+------------------------------+
| xenial | f4c4c60a6b75 | no     | ubuntu 16.04 LTS amd64 (release) (20160420.3) | x86_64 | 137.54MB | Jul 7, 2016 at 10:38pm (UTC) |
+--------+--------------+--------+-----------------------------------------------+--------+----------+------------------------------+
|        | 2cae69bd55ff | no     |                                               | x86_64 | 225.68MB | Jul 13, 2016 at 7:32pm (UTC) |
+--------+--------------+--------+-----------------------------------------------+--------+----------+------------------------------+
```

**Comment by stgraber** [comment]
Can you still reproduce this?

I tried here for good measure and can't reproduce the issue, so there's probably some environment difference or something was wrong with your system somehow...

**Comment by deskjet** [comment]
I'm not sure if this is related but I received the same error message when trying to publish a snapshot that was slightly more than 10 GB in size. After freeing up some space in the container publish works again. LXD is set up using LVM storage and the container has a 30 GB disk.

I also observed a new dead symlink in `/var/lib/lxd/images` aber publish failed.

The system is an up-to-date Ubuntu 16.04 (lxd 2.0.3). This is the output of `lxc info`:

```
apicompat: 0
auth: trusted
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  driver: lxc
  driverversion: 2.0.3
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 4.4.0-34-generic
  server: lxd
  serverpid: 3083
  serverversion: 2.0.3
  storage: lvm
  storageversion: 2.02.133(2) (2015-10-30) / 1.02.110 (2015-10-30) / 4.34.0
config:
  storage.lvm_vg_name: lxc
public: false

```

**Comment by stgraber** [comment]
Oh, right, so that's one case where this kind of error could happen.

In order to generate the image, LXD needs to make a tarball of the container, then compress it. So you need whatever disk /var/lib/lxd/images is on to have about 1.5x the size of the container in free space.

That's so the uncompressed tarball can be generated and then for there to be enough room to store the compressed version (uncompressed is then removed).

**Comment by deskjet** [comment]
I could reproduce the problem in a fresh Ubuntu Xenial VM with enough storage space for `/var/lib/lxd/images`. To reproduce the `exit status 2` message I did the following:

- install ubuntu xenial server on a regular primary partition (~30 GB)
- update system and install lxd, lxd-client, lvm2, thin-provisioning-tools
- reboot
- setup LVM volume group using another partition (~25 GB) , configure LXD to use it
- create a container (i used the ubuntu:xenial image)
- make container disk larger (`lvresize --resizefs --size +20GB /dev/lxc/disk-image`)
- create large file in container (`dd if=/dev/zero of=largefile bs=1G count=10`)
- snapshot container
- publish snapshot
- `exit status 2` after some time

The 30 GB host file system should be enough to create a 10 GB image even if 15 GB are needed to create the tar.gz. I also tried the same with to small storage for `/var/lib/lxd/images`. This resulted in a different error message about disk mentioning a disk space problem.

**Comment by stgraber** [comment]
Attempting to reproduce this now on Ubuntu 16.04 with LVM as described above.

**Comment by stgraber** [comment]
Right, so I managed to reproduce this. It looks like the tarball image gets created successfully. The step that fails is when the tarball is then extracted as a new filesystem.

I have a theory for why this fails. Basically we always create the LVs at 10GB or whatever value is set through configuration. If that value is smaller than the uncompressed size of the tarball, extract fails by running out of space inside the LV.

So the problem is what to do in this case. We don't know the real uncompressed size of an image when importing it, so we can't just allocate a larger LV when the image is bigger than the default LV size.
And even if we could, that'd just lead to any container created from such an image to immediately run out of disk space.

I think that the best we can do is catch that uncompress error, check if we ran out of space in the LV and if we did, print a clear error, suggesting to the user that they bump core.lvm_volume_size to something larger and try again.

This is obviously not ideal, but there's not much else we can do. There's no reliable way to know how big a compressed image will be once unpacked as it depends on the compression algorithm and format, and unfortunately most of them don't reliably export the uncompressed size...

I'll implement this "fix" tomorrow.
